dismantle-tablegen/src/Dismantle/Tablegen/BitTrie.hs

```diff
@@ -0,0 +1,513 @@
+based on byte values at each level of the tree, it dispatches one bit at a time.
+
+At each level of the trie construction, we have a set of patterns that matched
+up to the current bit.  For each concrete bit (True or False) at the current
+position, we filter the set of patterns down to those that match accounting for
+the current concrete bit.  We apply negative patterns as soon as possible.  Once
+we reach the end of the bit stream, we produce a leaf that parses using the
+single remaining matching pattern.
+
+There is a special case to handle some overlapping patterns where we take the
+'most specific pattern'; see makePayload for details.
+
+Some important semantic notes for the table construction process:
+
+- Negative patterns can only be matched once all of their required bits are
+  present.  It would probably be best to match them as soon as possible to
+  simplify the logic and reduce the size of matched pattern sets.
+
+- If table results are cached, the cache must respect negative patterns.  If we
+  don't respect negative patterns, there will be invalid patterns in the suffix
+  returned by the cache lookup
+
+  Instead of invalidating the cache at each lookup, it might be more efficient
+  to compute a predicate for cached entries that returns 'True' if the entry is
+  not valid for the "current"
+
+- We can match a negative pattern once we have reached the last set bit in the
+  negative pattern mask
+
+-}
+
+{- Note [Caching With Negation]
+
+The underlying problem with the cache is that we are keying it on the set of
+patterns that match up to this point.  If we reuse cached results on a bit
+sequence that triggers a negative pattern, we will get incorrect parse tables
+(i.e., they would ignore the negative patterns).  Precise caching would be
+*very* difficult.  Instead, we inspect the bit pattern to see if we are on a
+path that *could* trigger a negative pattern from the currently active pattern
+set.  If we are, we simply do not use the cached result.
+
+The most common use of negative patterns is to say that some N-bit operand is a
+register operand, but must NOT equal a certain value in order to be a valid
+parse (because the excluded value parses as another instruction).
+
+- Under this example, we *might* trigger the negative pattern at any time until
+  the bit pattern is fully resolved (i.e., parsing is past the last bit in the
+  negative pattern)
+- However, not caching at all up to that point is extremely expensive
+
+Question 1: Can we use cached values if the prefix does not include any of the
+bits in the negative mask?
+
+Question 2: Do we need to worry about any negative bits except for the last one?
+
+Observation: construction is DFS.  In a bit pattern like
+
+> P = |1|0|?|?|1|?|?|1|
+>      0 1 2 3 4 5 6 7
+
+where the second operand is not allowed to be 0b11.  We traverse out to the end
+of the bit-string depth first covering suffixes (starting from bit 5):
+
+- 0b001   # operand is 0b00 (and fine)
```

Comment:
This is still confusing (re my previous note).  Is `0b001` indicating bit 7 is set and bits 5 and 6 are zero?  Or is bit 5 set and bits 6 and 7 zero?  Or are these bits 0-2?